### PR TITLE
Update bundled Node to v12, which matches Atom's Electron 5

### PR DESCRIPTION
**Pull request by DeeDeeG**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### Description of the Change

- Update `keytar` to a version that builds on Node v12.
- Update Node to v12.0.0 to match Atom (which has Electron 5.0.13 --> Node 12.0.0).

<details><summary>Nitty-gritty details (click to expand):</summary>

- Atom has Electron 5.0.13
  - https://github.com/atom/atom/pull/20172 (PR is titled incorrectly. It's 5.0._13_, not _12_.)
  - See: ["electronVersion" in `package.json`](https://github.com/atom/atom/blob/02aa6b1c250d54a7a00a74a9592ce466d42afe5f/package.json#L15)
- Electron 5.x has Node 12.0.0, sort of.
  - Changelog entry for Electron 5.0.0 mentions Node 12.0.0: https://www.electronjs.org/releases/stable?version=5&page=3#5.0.0 
  - See: ['node_version' in DEPS at the electron repo](https://github.com/electron/electron/blob/5-0-x/DEPS#L14-L15)
    - [The Node commit Electron 5.x uses](https://github.com/nodejs/node/commit/666c67e078bddc32f25409f4a929c1e9b5f47373) is "approximately Node 12", but a couple hundred commits behind the actual stable [`v12.0.0` tag](https://github.com/nodejs/node/tree/v12.0.0) at `github.com/nodejs/node`???
      - Mentioned here by electron maintainers: https://github.com/electron/electron/pull/19292#discussion_r304142155
      - Comparison of `666c67e` to `v12.0.0` at nodejs/node: https://github.com/nodejs/node/compare/666c67e078bddc32f25409f4a929c1e9b5f47373...v12.0.0

Latest Atom nightly reports Node v12.0.0:

```console
$ atom-nightly  --version
Atom    : 1.49.0-nightly0
Electron: 5.0.13
Chrome  : 73.0.3683.121
Node    : 12.0.0
```

</details>

### Alternate Designs

- Try to get Atom to update to a newer version of Electron (they are kind of overdue).
  - We would still be updating to some Node v12.x release. Just a somewhat newer minor/patch.
    - Which Node is used by which Electron? See: https://www.electronjs.org/docs/tutorial/electron-timelines

Or

- YOLO: use some newer version of Node v12.x that doesn't strictly match Atom.

Node major versions are meant to be fairly stable, so in theory we can use any version of Node v12.x (not just strictly the first release in the series, v12.0.0). But this is already an area where the failures are obscure/hard to track down. The more we differ here from the node that will run Atom packages, the wider the opening for behavior differences --> bugs. So to err on the safe side, we can match Electron as closely as possible. I just went with Node v12.0.0 to match Atom/Electron as closely as possible.

### Benefits

Should handle some certain cases where Atom packages/their dependencies (various node modules) need to be built and run on the same version of Node. For example: `node-sass` as discussed in https://github.com/atom/apm/issues/328.

### Possible Drawbacks

As always, major version bumps are a thing to watch for breakages.

While this is newer than the Node we had for a while, v10.2.1, it is older than what we just upgraded to last, v10.20.1. Any bugs fixed in Node 12.x in the past year aren't included in this. (Probably just another reason to bug the Atom team to update to newer Electron, IMO.)

- https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V12.md#12.0.0
- https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V10.md#10.2.1
- https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V10.md#10.20.1

### Verification Process

I ran the following: `./bin/npm test` and `./bin/apm install tree-view`. CI should pass.

### Applicable Issues

See [past](https://github.com/atom/apm/pull/831) [node](https://github.com/atom/apm/pull/796) [upgrades](https://github.com/atom/apm/pull/694) [in this repo](https://github.com/atom/apm/pull/563) for relevant discussion.

### CODE REVIEW

**Comment by DeeDeeG** [comment]
Test failure is unrelated to changes. One of the runs on AppVeyor had some network issue, apparently, because it failed to reach github.com to download the repo/PR.

**Comment by DeeDeeG** [comment]
Atom has a PR open to update to Electron v6.1.12, so I think this PR should wait to see where that goes.

**Comment by DeeDeeG** [comment]
The electron 6 PR is about 20 days old, with as far as I can tell a major performance regression on Linux, so I'm not sure when it will merge. This PR, as it is, would match Atom `master` with Electron 5.0.13.

**Comment by DeeDeeG** [comment]
cc @smashwilson @lkashef @darangi 

Sorry to bother you all.

This is a PR to sync Node in `apm` with the version used by Electron 5.0.13 in Atom. i.e. it should be Node v12.0.0.

Please consider:
- merging this PR at `apm`
- then bringing in the updated `apm` at Atom here: https://github.com/atom/atom/tree/master/apm

Thank you!

\- DeeDeeG

**Comment by lkashef** [comment]
@DeeDeeG the drawback seems like a big risk, I would be more comfortable doing this locally, making sure that all atom's node modules are compatible then I would upgrade

Let me know if I missed anything

**Comment by DeeDeeG** [comment]
The packages we want to be able to build to address the "downside" are these: https://atom.io/packages (and of those, only ones with native code, generally C++, which is somewhat uncommon in the Node ecosystem, in favor of pure JS, but certainly not unheard-of).

I can try to do `apm install [package from the atom.io package registry]` for many packages against Node 12, and see if it all works.

I can further take notes, and report somewhere which packages don't build. (And we can do a courtesy issue at those packages to alert the maintainers they might want to test against Node 12? Give them some maybe pre-defined lead-time so the package ecosystem has time to respond?)

---

Limited scope of potential impact: I recall from before that most packages have no native code anyhow. (Only packages with native code need to be "built" -- all other NodeJS code is simply "interpreted" live in the current Electron/Node/V8 runtime environment) These were the ones I could find last time, maybe I'll start with those: https://github.com/atom/atom/pull/20406#issuecomment-616162951

<details><summary>List of packages I've found with native code to build (click to expand):</summary>


    https://atom.io/packages/language-rust-treesitter
    https://atom.io/packages/tree-sitter-4dm
    https://atom.io/packages/language-python-treesitter
    https://atom.io/packages/language-zig-treesitter
    https://atom.io/packages/language-tree-test
    https://atom.io/packages/language-lux
    https://atom.io/packages/language-janet
    https://atom.io/packages/language-carp
    https://atom.io/packages/greyatom-hydrogen
    https://atom.io/packages/atom-ide-terminal
    https://atom.io/packages/terminal-tab-plus
    https://atom.io/packages/terminal-tab-service
    https://atom.io/packages/electric-terminal
    https://atom.io/packages/ls-compile
    https://atom.io/packages/build-haxe
    https://atom.io/packages/flatpak-builder-runner

</details>

---

In summary: Atom packages with native code should be tested with apm using Node 12, in order to get a sense of the impact.

**Comment by DeeDeeG** [comment]
> I would be more comfortable doing this locally, making sure that all atom's node modules are compatible then I would upgrade

If I recall correctly, the packages bundled in Atom needed to be buildable with Node 12 in order for the Electron 5.0.13 update to happen at https://github.com/atom/atom. I think those packages are ready as part of https://github.com/atom/atom/pull/20172.

I also seem to recall building atom with `apm` from this git branch and it worked. I will post a CI run of that (from my forked repo of Atom) if I get the time to make that happen.

**Comment by DeeDeeG** [comment]
Updating as I test:

- [ ] https://atom.io/packages/language-rust-treesitter
- [ ] https://atom.io/packages/tree-sitter-4dm
- [ ] https://atom.io/packages/language-python-treesitter
- [ ] https://atom.io/packages/language-zig-treesitter
- [ ] https://atom.io/packages/language-tree-test
- [ ] https://atom.io/packages/language-lux
- [ ] https://atom.io/packages/language-janet
- [ ] https://atom.io/packages/language-carp
- [ ] https://atom.io/packages/greyatom-hydrogen
- [ ] https://atom.io/packages/atom-ide-terminal
- [ ] https://atom.io/packages/terminal-tab-plus
- [ ] https://atom.io/packages/terminal-tab-service
- [ ] https://atom.io/packages/electric-terminal
- [ ] https://atom.io/packages/ls-compile
- [ ] https://atom.io/packages/build-haxe
- [ ] https://atom.io/packages/flatpak-builder-runner

check mark means it installed with no problems.

#### Steps to reproduce successful package installs with Node 12.0.0:

1. Install the official stable Atom release from here: https://atom.io/
2. Clone the `apm` git repo, get my branch somehow, and checkout this `node-12` branch
3. `npm install` to get `apm` ready to use
4. Still in this repo, do `./bin/apm install [some-package-here]`
   1. To see `node-gyp` rebuilding the native code, run with `--verbose`, i.e. `./bin/apm install --verbose [some-package-here]`

##### Recommended way to get my branch locally:

1. Clone this repo `git clone https://github.com/atom/apm`
1. `cd apm`
1. Add my fork as a remote `git remote add deedeeg https://github.com/DeeDeeG/apm`
1. Fetch the content of my fork: `git remote update deedeeg`
1. Switch to my branch: `git switch node-12`
   1. If that doesn't work, more explicitly: `git switch -c deedeeg-node-12 deedeeg/node-12` (creates a branch `deedeeg-node-12` tracking the remote branch `node-12` from my fork)
1. Run `npm install` to get `apm` ready to use
1. Now finally you can do `./bin/apm install [some-package-here]`

**Comment by DeeDeeG** [comment]
Two packages that I tried errored out, but they don't build against `apm` at `master` branch with Node 10 either.

One of them has this in the README.md: 
> _**THIS IS PRE-PRE-ALPHA -- BUYER BEWARE**_

(from: https://github.com/futuretim/flatpak-builder-runner#readme)

And the other package has "test" in the name, (https://github.com/Aerijo/language-tree-test) so I'm not sure it was ever meant for wide use. The author is active around Atom, so maybe they can let us know if they have any issues with this PR? `@Aerijo`.

But I genuinely think the package doesn't work, at least not against Atom stable v49 with Electron 5.

**Comment by DeeDeeG** [comment]
On second thought, we are always going to ultimately build against Electron. See: https://github.com/atom/apm/blob/adb72dc9ca7563f060f00d0f652eb4e641b852d8/src/command.coffee#L109-L113).

Docs for these environment variables, which we set to build against Electron, not Node: https://github.com/electron/electron/blob/master/docs/tutorial/using-native-node-modules.md#using-npm

This is just to benefit those select few packages that directly read the current running Node (presumably from environment variables) in a way that isn't reliable in our case... A thing we can do to help those users is to bundle the same Node version as our bundled Electron uses.

I don't expect any Atom package would build/not build based on this change alone. I may have been mistaken when listing that as a "downside" for this PR.

**Comment by DeeDeeG** [comment]
Comments above are updated. I can confirm this does NOT affect if community packages can be installed to Atom.

**Comment by DeeDeeG** [comment]
I have a passing CI run of Atom built with `apm` from this PR (with bundled Node 12).

https://dev.azure.com/DeeDeeG/b/_build/results?buildId=125&view=results

There are assets attached if you want to try it. If you install this and do `apm-dev install --verbose [any-package-here]` you can see it working. Particularly if the package has native C++ code to build. And yes, this was able to build Atom's internal packages, no problem.

**Comment by Aerijo** [comment]
> And the other package has "test" in the name, (https://github.com/Aerijo/language-tree-test) so I'm not sure it was ever meant for wide use. The author is active around Atom, so maybe they can let us know if they have any issues with this PR? @Aerijo.

@DeeDeeG The test refers to being for Tree-sitter test files. I can update it to work though, should just be a matter of maybe bumping the Tree-sitter dependencies and recompiling.

**Comment by aminya** [comment]
@DeeDeeG Could you hide the outdated comments. I am a little confused here.

**Comment by DeeDeeG** [comment]
Summary:

`apm` doesn't really need any particular Node except for in rare cases when a package "sniffs" for what Node is being run and makes some decisions based on that. apm builds packages against Electron, not Node, for all the actual V8 headers and Node API stuff.

Updating this shoukd solves those rare cases, in theory, and as my past few comments have laid out the evidence,mI can confidently say there is no downside (hardly any imoact at all really?) of updating the Node bundled with `apm`.

(I'm not hiding those comments at the moment, because they establish the facts to support my claims that this PR does not cause problems.)

(I wonder if apm could get by without bundling its own Node, but until someone figures out how to do that, this PR/bumping Node to match Atom's Electron is still relevant for those few modules that detect the running Node at install time.)

**Comment by DeeDeeG** [comment]
@Aerijo after doing some more research, this PR seems to be unrelated after all, with regard to if your package builds or not. (Or any package, for that matter). It's more about preventing rare bugs like https://github.com/atom/apm/issues/328. And because it's "right," I suppose, to match Electron.

**Comment by DeeDeeG** [comment]
@lkashef 

I was able to confirm the "downside" of this PR doesn't exist. I was just confused about how `apm` works when I wrote that "potential downsides" section. (Sorry about that.)

Briefly: Atom builds packages against Electron. Atom ships with Electron 5.0.13, which has Node 12.0.0. All packages wanting to be installed to Atom _already_ need to build against Node 12, as of Atom 1.47.0. https://github.com/atom/atom/releases/tag/v1.47.0

This is confirmed by my testing above. [[1]](https://github.com/atom/apm/pull/889#issuecomment-660266507) [[2]](https://github.com/atom/apm/pull/889#issuecomment-660277973) This PR doesn't cause build failures for any packages I tested.

As mentioned in prior comments, what effect does the PR have then? Strictly limited to fixing some obscure bugs in node ecosystem packages if they see the wrong "Node version" in the environment at install time. (For example: https://github.com/atom/apm/issues/328)

I want to apologize for any confusion or concern I caused, as it was based on a misunderstanding of mine.

File: BUNDLED_NODE_VERSION [hunk 1]
```diff
@@ -1 +1 @@
-v10.20.1
+v12.0.0
```
Comment (aminya) [inline BUNDLED_NODE_VERSION:1]: For upgrading to Electron 7, we should update this to at least 12.13.1.
NSFW needs this for compiling: https://github.com/atom-ide-community/atom/pull/75

**Comment by aminya** [comment]
@lkashef @darangi I am facing the issues without this on Electron 6. Native modules that are compiled for the old apm-bundled Node will not work on newer Electron. This is absolutely necessary (treat it like an urgent patch). The Node versions for Atom and APM should be identical to avoid these problems.

I suggest that you merge this before upgrading the electron on Atom. Then in another PR, the Node version should be bumped again to match Electron 6.


This is the error I am getting trying to run `ink` and `julia-client`. As you see the Node.js versions is making problems here.
```
The module '\\?\D:\GitHub\JavaScript\@junolab\atom-julia-client\node_modules\node-pty-prebuilt-multiarch\build\Release\pty.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 70. This version of Node.js requires
NODE_MODULE_VERSION 73. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
    at process.func (electron/js2c/asar.js:138:31)
    at process.func [as dlopen] (electron/js2c/asar.js:138:31)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:828:18)
    at Object.func (electron/js2c/asar.js:138:31)
    at Object.func [as .node] (electron/js2c/asar.js:138:31)
    at Module.load (internal/modules/cjs/loader.js:645:32)
    at Function.Module._load (internal/modules/cjs/loader.js:560:12)
    at Module.require (file:///D:/GitHub/JavaScript/atom-dev/resources/app.asar/static/index.js:72:46)
    at require (D:\GitHub\JavaScript\atom-dev\resources\app\static\<embedded>:11:146814)
    at new WindowsPtyAgent (D:\GitHub\JavaScript\@junolab\atom-julia-client\node_modules\node-pty-prebuilt-multiarch\lib\windowsPtyAgent.js:52:36)
    at new WindowsTerminal (D:\GitHub\JavaScript\@junolab\atom-julia-client\node_modules\node-pty-prebuilt-multiarch\lib\windowsTerminal.js:49:24)
    at Object.spawn (D:\GitHub\JavaScript\@junolab\atom-julia-client\node_modules\node-pty-prebuilt-multiarch\lib\index.js:28:12)
    at file:///C:/Users/yahyaaba/Documents/GitHub/JavaScript/github/julia-client/lib/connection/process/basic.js:58:24
    at processTicksAndRejections (internal/process/task_queues.js:89:5)
```

**Comment by DeeDeeG** [comment]
@aminya I see you ran `apm rebuild`, which should mean this is a real issue...

---

The strange thing is, those `NODE_MODULE_VERSION` numbers aren't Node ABI versions, they are Electron ABI versions: https://github.com/lgeiger/node-abi/blob/75cc2ddc7f4d978a8d44a64877f2edb17b555189/index.js#L74-L75

(Bonus link: https://github.com/nodejs/node/blob/dd0c5228acaf5695f96809e50771ead3167b0f44/doc/abi_version_registry.json#L16-L19)

So, these packages are being built against Electron 5.x, yet the Electron app (Atom) trying to run them is expecting Electron 6.x. There should be a UI button somewhere in Atom with a bug (ladybug) icon and a little red x in a circle (if I recall correctly) that when you click on it shows an error message, which mentions the packages need to be rebuilt, and you can click "rebuild packages." Does this solve the problem?

<details><summary>Here's the table of Node ABI numbers (`NODE_MODULE_VERSION`)</summary>

https://nodejs.org/en/download/releases/

Electron and Node have separate ABI numbers, see: .
</details>

**Comment by aminya** [comment]
> when you click on it shows an error message, which mentions the packages need to be rebuilt, and you can click "rebuild packages." Does this solve the problem?

This is not shown! 

You can reproduce the error by installing https://atom.io/packages/ink

**Comment by DeeDeeG** [comment]
I can't reproduce this problem. The package uninstalls and reinstalls fine.

Screenshot showing Atom-dev built from commit 6c2fed8, with Electron 6, having the ink package installed in the settings tab

My install of Atom-dev is built from the [CI artifact "atom-x64-windows.zip"](https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/build/builds/443/artifacts?artifactName=atom-x64-windows.zip&api-version=5.1&%24format=zip) from our `master` branch.

<details><summary>Here is my script to build this into an installer (must extract CI artifact zip contents to `atom\out\Atom Dev x64`, and script must be saved in and run from `atom\script\`):</summary>

```js
// atom\script\create-windows-installer.js

const path = require('path');

const CONFIG = require('./config');
const createWindowsInstaller = require('./lib/create-windows-installer');

packagedAppPath = path.join(CONFIG.buildOutputPath, CONFIG.appName);
if (process.platform === 'win32' && process.arch !== 'ia32') {
  packagedAppPath += ` ${process.arch}`;
}

createWindowsInstaller(packagedAppPath);
```

</details>

**Comment by aminya** [comment]
The installation is fine, but you can't run the REPL. You need `julia` on your path and you need to have `julia-client` package installed too. 

Annotation 2020-08-12 001702

**Comment by DeeDeeG** [comment]
Thanks for the updated info. I personally still can't reproduce this.

Screenshot showing Atom-dev with Electron 6, and using the Julia REPL to print "hello world"

**Comment by DeeDeeG** [comment]
I was able to reproduce the failure to start the REPL when switching back and forth between Atom with different Electron version. 

Steps To Reproduce:
- Install `julia` language binary and put it on the system PATH somewhere
- Start a version of Atom
- Install `julia-client` package
  - This asks to install a dependency `language-julia`, accept this and install the dependency
- Note that REPL is working
- Quit Atom, and start the other version of Atom with a different major Electron version
- Note that julia REPL isn't working (failure appears as shown in @aminya's screenshots)

I also found a solution that's working for me at the moment.

Steps to fix:
- Open the Atom version you intend to use with Julia
- In Settings, uninstall and reinstall both the `language-julia` and `julia-client` packages
- May have to quit and restart Atom for REPL to be working again, I'm not 100% sure
- Julia REPL in Atom should be working again.

It's odd that Atom doesn't prompt to rebuild stale packages, but oh well.

**Comment by aminya** [comment]
The problem is with the code that detects the incompatible package. The cache probably returns an incorrect result. 

https://github.com/atom/atom/blob/2d3e332b889bce1e458d154efa608016f47d8b9d/src/package.js#L1283

**Comment by DeeDeeG** [comment]
I note that `julia-client` uses `node-pty-prebuilt-multiarch`and spends a suspicious amount of time with an active node subprocess after it looks like installation finishes if I do `apm-dev install julia-client --verbose`. I'm not sure why the install doesn't exit quickly. It makes me think the (unusual?) way the package is installed defeats the usual `getIncompatibleNativeModules()` check.

In any case, this discussion is technically off-topic for the present issue. If we find out more, we can post a good quality issue to Atom with all the necessary information to work on a fix.

**Comment by DeeDeeG** [comment]
One last off-topic comment, sorry.

I found that the stale module detection worked just fine on my Ubuntu 20.04 Linux machine:

Atom showing an error message, which says that julia-client could not be loaded because it contains "native modules that were compiled for an earlier version of Atom."

**Comment by darangi** [comment]
Thanks for the contribution @DeeDeeG 🙇🏾‍♂️

**Comment by DeeDeeG** [comment]
Thank you for merging this @darangi!

This PR bumps the bundled Node to _**v12.0.0**_. I notice this is in the Electron 6 PR now; For Electron 6, bundled Node should be **_v12.4.0._**

**Comment by darangi** [comment]
Thanks for the heads up. I will open a PR to fix that.
